### PR TITLE
Support functions, but without tail recursion

### PR DESCRIPTION
**Pull request by anubh-v**
With this PR:
- We can declare functions, but tail recursive calls are not optimised
- We can apply functions that we've declared OR the built in functions

Try in the REPL:
`list(amb(1,2,3));`

`function integer_from(n) { amb(n, integer_from(n + 1)); } integer_from(1);`

```
function an_element_of(xs) { require(!is_null(xs)); amb(head(xs), an_element_of(tail(xs))); } an_element_of(list(1,2,3));
```

### CODE REVIEW

File: src/__tests__/non-det-interpreter.ts [hunk 1]
```diff
@@ -12,9 +12,24 @@ test('Deterministic calculation', async () => {
...
-test('Deterministic assignment', async () => {
-  await testDeterministicCode('let a = 5; a = 10; a;', 10)
+test('Test builtin list functions', async () => {
+  await testDeterministicCode('pair(false, 10);', [false, 10])
+  await testDeterministicCode('list();', null)
+  await testDeterministicCode('list(1);', [1, null])
+  await testDeterministicCode('head(list(1));', 1)
+  await testDeterministicCode('tail(list(1));', null)
+  await testDeterministicCode(
+    `function increment(x) {
```
Comment (arsalan0c) [inline src/__tests__/non-det-interpreter.ts]: This test fails. It seems it is because `map` uses the `return` statement.

**Review by arsalan0c — CHANGES_REQUESTED** [review]
It would be good to add tests for sub-functions as well as functions with `amb` arguments to ensure the Cartesian product gives the expected behaviour

(Tests added in #11)

File: src/__tests__/non-det-interpreter.ts (hunk 1 above)
Comment (anubh-v) [reply src/__tests__/non-det-interpreter.ts]: Yea, I realised that `map` is a prelude function (written in Source) while others like `pair` / `head` are "builtins" (written in Typescript).

i.e. the prelude functions will only work when we support return statements

**Comment by anubh-v** [comment]
I've added a very basic *draft* of the return statements' support (without tail call optimisation)

**Review by arsalan0c — APPROVED** [review]

**Pull request merged**